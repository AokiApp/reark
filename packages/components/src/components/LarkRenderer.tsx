import React, { useContext } from "react";
import { LarkApiContext } from "../contexts/LarkApiContext";
import { BlockStoreProvider } from "../contexts/BlockStoreContext";
import { BlockComponent } from "./BlockComponent";

export const LarkRenderer: React.FC = () => {
  const { blocks } = useContext(LarkApiContext);

  if (!blocks || blocks.length === 0) return null;

  const rootBlock =
    blocks.find((block) => block.block_type === 1 && !block.parent_id) ||
    blocks[0];

  if (!rootBlock) return null;

  return (
    <BlockStoreProvider items={blocks}>
      <BlockComponent blockId={rootBlock.block_id} />
    </BlockStoreProvider>
  );
};
