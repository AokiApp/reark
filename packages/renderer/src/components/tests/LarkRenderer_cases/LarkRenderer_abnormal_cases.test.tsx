import { renderWithVRT } from "../test-utils/renderWithVRT";
import { LarkRenderer } from "../../LarkRenderer";

// LarkRenderer 異常系: 空データ
describe("LarkRenderer 異常系: 空データ", () => {
  it("blocksが空配列の場合はnullを返す", async () => {
    const { container, vrt } = renderWithVRT(
      <LarkRenderer initialData={{ blocks: [] }} />,
    );
    await vrt();
    expect(container.firstChild).toBeNull();
  });

  it("blocksがundefinedの場合はnullを返す", async () => {
    const { container, vrt } = renderWithVRT(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      <LarkRenderer initialData={{ blocks: undefined as any }} />,
    );
    await vrt();
    expect(container.firstChild).toBeNull();
  });
});

// LarkRenderer 異常系: 不正なblock_type
describe("LarkRenderer 異常系: 不正なblock_type", () => {
  it("未知のblock_typeを含む場合もUnsupportedBlockとしてレンダリングされる", async () => {
    const block = {
      block_id: "invalid-block",
      block_type: 9999,
      parent_id: "",
      children: [],
      text: "不正なブロック",
    };
    const blocks = [block];
    const { container, vrt } = renderWithVRT(
      <LarkRenderer initialData={{ blocks }} />,
    );
    await vrt();
    expect(container).toMatchSnapshot();
  });
});

// LarkRenderer 異常系: childrenプロパティ欠損
describe("LarkRenderer 異常系: childrenプロパティ欠損", () => {
  it("childrenプロパティが存在しないブロックもレンダリングできる", async () => {
    const block = {
      block_id: "no-children",
      block_type: 2,
      parent_id: "",
      text: {
        elements: [
          {
            text_run: {
              content: "children欠損",
              text_element_style: { bold: false },
            },
          },
        ],
        style: { align: "left", folded: false },
      },
      // childrenプロパティ無し
    };
    const blocks = [block];
    const { container, vrt } = renderWithVRT(
      <LarkRenderer initialData={{ blocks }} />,
    );
    await vrt();
    expect(container).toMatchSnapshot();
  });
});

// LarkRenderer 異常系: childrenが空配列
describe("LarkRenderer 異常系: childrenが空配列", () => {
  it("childrenが空配列でも正常にレンダリングできる", async () => {
    const block = {
      block_id: "empty-children",
      block_type: 2,
      parent_id: "",
      text: {
        elements: [
          {
            text_run: {
              content: "children空配列",
              text_element_style: { bold: false },
            },
          },
        ],
        style: { align: "left", folded: false },
      },
      children: [],
    };
    const blocks = [block];
    const { container, vrt } = renderWithVRT(
      <LarkRenderer initialData={{ blocks }} />,
    );
    await vrt();
    expect(container).toMatchSnapshot();
  });
});

// LarkRenderer 異常系: 親無し子ブロック
describe("LarkRenderer 異常系: 親無し子ブロック", () => {
  it("親が存在しないchildrenを持つブロックもレンダリングできる", async () => {
    const childBlock = {
      block_id: "orphan-child",
      block_type: 2,
      parent_id: "non-existent-parent",
      text: {
        elements: [
          {
            text_run: {
              content: "親無し子ブロック",
              text_element_style: { bold: false },
            },
          },
        ],
        style: { align: "left", folded: false },
      },
      children: [],
    };
    const blocks = [childBlock];
    const { container, vrt } = renderWithVRT(
      <LarkRenderer initialData={{ blocks }} />,
    );
    await vrt();
    expect(container).toMatchSnapshot();
  });
});
