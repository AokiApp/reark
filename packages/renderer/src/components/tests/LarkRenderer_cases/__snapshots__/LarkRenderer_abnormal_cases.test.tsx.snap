// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LarkRenderer 異常系: childrenが空配列 childrenが空配列でも正常にレンダリングできる 2`] = `
<div>
  <div
    class="reark-block"
  >
    <div
      class="reark-text-block"
      style="text-align: left;"
    >
      <span
        class="reark-text"
      >
        children空配列
      </span>
    </div>
  </div>
</div>
`;

exports[`LarkRenderer 異常系: childrenプロパティ欠損 childrenプロパティが存在しないブロックもレンダリングできる 2`] = `
<div>
  <div
    class="reark-block"
  >
    <div
      class="reark-text-block"
      style="text-align: left;"
    >
      <span
        class="reark-text"
      >
        children欠損
      </span>
    </div>
  </div>
</div>
`;

exports[`LarkRenderer 異常系: 不正なblock_type 未知のblock_typeを含む場合もUnsupportedBlockとしてレンダリングされる 2`] = `
<div>
  <div
    class="reark-unsupported-block"
  >
    <strong>
      Unsupported block type: 
      9999
    </strong>
  </div>
</div>
`;

exports[`LarkRenderer 異常系: 親無し子ブロック 親が存在しないchildrenを持つブロックもレンダリングできる 2`] = `
<div>
  <div
    class="reark-block"
  >
    <div
      class="reark-text-block"
      style="text-align: left;"
    >
      <span
        class="reark-text"
      >
        親無し子ブロック
      </span>
    </div>
  </div>
</div>
`;
